### PR TITLE
add useLocalProxy react hook

### PR DESCRIPTION
**Pull request by steve8708**
via https://codesandbox.io/s/hopeful-merkle-ziu1h?file=/src/App.js:105-278

only change I made was added TS typings and allow the argument to be a function in case intializing the state object is non trivial and you don't want that running every render when only needed on first render

### CODE REVIEW

**Comment by dai-shi** [comment]
Hi, I was thinking about adding useLocalProxy too. But, not in the core. I would put it in utils because we want to keep the core api minimal and it's just a wrapper function around the core.

```js
import { useLocalProxy } from 'valtio/utils';
```

I can merge this, and continue the work on it on my end, if you are ok with it.

**Comment by steve8708** [comment]
Thanks @dai-shi - makes sense and that works for me, and happy to update JSX Lite to use `import { useLocalProxy } from 'valtio/utils';` or whatever the final API is when ready 👍

**Comment by dai-shi** [comment]
Sounds good. I think I can add commits to this branch. Let me try.

File: src/index.ts [hunk 1]
```diff
@@ -53,4 +53,13 @@ const useProxy = <T extends object>(p: T): T => {
   return createDeepProxy(currSnapshot, affected, proxyCache)
 }
 
-export { proxy, subscribe, snapshot, useProxy }
+const useLocalProxy = <T extends object>(init: T | (() => T)) => {
+  const ref = useRef<T>()
+  if (!ref.current) {
+    const initialObject = typeof init === 'function' ? (init as any)() : init
```
Comment (Aslemammad) [inline src/index.ts:59]: Just for the suggestion, Can we use `(init as () => T)` instead of `(init as any)`?

**Review by Aslemammad — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 1 above)
Comment (steve8708) [reply src/index.ts]: ah yes ofc - I think @dai-shi is going to take over this PR but went ahead and threw that update in bc I agree `any` is ugly

**Comment by dai-shi** [comment]
Created a simple example: https://codesandbox.io/s/react-typescript-forked-qye5u?file=/src/App.tsx

File: readme.md [hunk 2]
```diff
@@ -98,3 +98,14 @@ subscribe(state, () => {
   const obj = snapshot(state) // A snapshot is an immutable object
 })
 ```
+
+##### Use it locally in components
+
+You can use it locally in components. [Note](./src/utils.ts#L5-L15)
+
```
Comment (Aslemammad) [inline readme.md:105]: Now it's great, but can't we put the notes here, I don't have any problem with this, but just a question.

**Review by Aslemammad — COMMENTED** [review]
I feel great about it, I put a comment in the readme.

File: readme.md (hunk 2 above)
Comment (dai-shi) [reply readme.md:105]: I just feel a bit long..

Comment (Aslemammad) [reply readme.md:105]: That's ok, no problem.

**Comment by dai-shi** [comment]
Let's merge this in.

**Comment by dai-shi** [comment]
https://www.npmjs.com/package/valtio/v/0.4.5

**Pull request merged**